editor/components/inserter/menu.js

```diff
@@ -29,32 +77,206 @@ class InserterMenu extends wp.element.Component {
 		return () => {
 			this.props.onInsertBlock( slug );
 			this.props.onSelect();
-			this.setState( { filterValue: '' } );
+			this.setState( {
+				filterValue: '',
+				currentFocus: null
+			} );
 		};
 	}
 
-	render() {
-		const { position = 'top' } = this.props;
-		const blocks = wp.blocks.getBlocks();
-		const isShownBlock = block => block.title.toLowerCase().indexOf( this.state.filterValue.toLowerCase() ) !== -1;
-		const blocksByCategory = blocks.reduce( ( groups, block ) => {
-			if ( ! isShownBlock( block ) ) {
-				return groups;
-			}
-			if ( ! groups[ block.category ] ) {
-				groups[ block.category ] = [];
+	getVisibleBlocks( blockTypes ) {
+		return blockTypes.filter( this.isShownBlock );
+	}
+
+	sortBlocksByCategory( blockTypes ) {
+		const getCategoryIndex = ( item ) => {
+			return this.categories.findIndex( ( category ) => category.slug === item.slug );
+		};
+
+		return blockTypes.sort( ( a, b ) => {
+			return getCategoryIndex( a ) - getCategoryIndex( b );
+		} );
+	}
+
+	groupByCategory( blockTypes ) {
+		return blockTypes.reduce( ( accumulator, block ) => {
+			// If already an array push block on else add array of block.
```

Comment:
We could simplify this somewhat by just creating the array if not exists and using the same `push` and `return` behavior.

```js
return blockTypes.reduce( ( accumulator, block ) => {
	// If already an array push block on else add array of block.
	if ( ! accumulator[ block.category ] ) {
		accumulator[ block.category ] = [];
	}

	return accumulator[ block.category ].concat( block );
}, {} );
```

Alternatively, maybe this could be more easily achieved with [Lodash's `_.groupBy`](https://lodash.com/docs/4.17.4#groupBy).